File: leetcode/binary_tree_construction/binary_tree_construction.js
var buildTree = function(preorder, inorder) {
  let inOrderMap = {}
  for (let i=0; i < inorder.length; i++) {
    inOrderMap[inorder[i]] = i
  }
  let root = constructTree(0, inorder.length-1, preorder, [0], inOrderMap)
  return root
};
const constructTree = (start, end, preorder, preOrderIndex, inOrderMap) => {
  if (start > end) return null
  const root = new TreeNode(preorder[preOrderIndex[0]++])
  const index = inOrderMap[root.val]
  // Construct left sub tree
  root.left = constructTree(start, index-1, preorder, preOrderIndex, inOrderMap)
  // Construct right sub tree
  root.right = constructTree(index+1, end, preorder, preOrderIndex, inOrderMap)
  return root
}
let preorder, inorder
preorder = [ 1, 2, 4, 3, 5, 7, 8, 6 ]
inorder = [ 4, 2, 1, 7, 5, 8, 3, 6 ]
console.log(buildTree(preorder, inorder))

File: leetcode/binary_tree_construction/binary_tree_construction.rb
# Definition for a binary tree node.
class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val)
        @val = val
        @left, @right = nil, nil
    end
end

# @param {Integer[]} preorder
# @param {Integer[]} inorder
# @return {TreeNode}
def build_tree(preorder, inorder)
  i = preorder.length - 1
  seq = 0
  rt = nil
  left = nil
  parent = nil
  root = TreeNode.new(preorder[0])
  while i > 0 do
    case seq
    when 0
      rt = TreeNode.new(preorder[i])
    when 1
      left = TreeNode.new(preorder[i])
    when 2
      parent = TreeNode.new(preorder[i])
      parent.right = rt
      parent.left = left
      if root.right.nil?
        # TODO: Need to handle larger trees...
        root.right = parent
      else
        root.left = parent
      end
      seq = 0
    else
      raise "Invalid seq #{seq}"
    end
    seq += 1
    i -= 1
  end
  return root
end

r = build_tree([1,2,4,5,3,6,7],nil)
p r.val
p r.left.val
p r.right.val

